### PR TITLE
feat(synthetic-shadow): innerText and outerText polyfills

### PR DESCRIPTION
**Pull request by jodarove**
## Details
This PR provides an `innerText` and `outerText` polyfill for synthetic-shadow, _behind a feature flag_.

The polyfill is based on the [innerText spec](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute), though it has some known drawbacks (so far):

- for the last `table-cell` within a `table-row`, an extra `\t` will be added.
- for the last `table-row` within a `table`, an extra `\n` will be added.
- the inner text value varies from browser to browser, ex: https://codepen.io/jodarove/pen/vYGmrXj is different in chrome and safari.

Pending: innerText setter implementation.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts [hunk 1]
```diff
@@ -116,6 +344,53 @@ function lastElementChildGetterPatched(this: ParentNode) {
     return children.item(children.length - 1) || null;
 }
 
+defineProperties(HTMLElement.prototype, {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts:347]: why is this here and not in html-element.ts file?

Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts:347]: missed it! good catch!

File: packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        it('should not collect text from hidden select', () => {
+            const testCase = elm.shadowRoot.querySelector('.select-hidden-testcase');
+
+            expect(testCase.innerText).toBe(`Chose a pet:`);
+        });
+
+        it('should not collect text from textarea', () => {
+            const testCase = elm.shadowRoot.querySelector('.textarea-testcase');
+
+            expect(testCase.innerText).toBe(`Tell us your story:`);
```
Comment (jodarove) [inline packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js:84]: in firefox, is displaying the text from the textarea: seems like the selection api can select the text node inside the text area.

File: packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            // 2. the last \n is incorrect, it's also a relaxation from the spec.
+            expect(testCase.innerText).toBe(`1,1\t1,2\t
+2,1\t2,2\t
+`);
+        });
+
+        it('should collect text from select', () => {
+            const testCase = elm.shadowRoot.querySelector('.select-testcase');
+
+            expect(testCase.innerText).toBe(`Chose a pet:
```
Comment (jodarove) [inline packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js:70]: in safari, `getComputedStyle(<option>) === 'inline'` while in other major browsers is block.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 4]
```diff
@@ -314,16 +324,16 @@ function innerTextPatched(this: Element): string {
     const thisComputedStyle = getElementComputedStyle(this);
     // 1. If this is not being rendered or if the user agent is a non-CSS user agent, then return this's descendant text content.
     if (!nodeIsBeingRendered(thisComputedStyle)) {
-        return textContentGetterPatched.call(this);
+        return this.textContent || ''; // textContentGetterPatched.call(this);
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts]: dead code

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 5]
```diff
@@ -255,9 +262,12 @@ function innerTextCollectionSteps(node: Node): InnerTextCollectionResult[] {
         } else if (tagName === 'TEXTAREA') {
             return [];
         } else {
-            node.childNodes.forEach((childNode) => {
+            forEach.call(node.childNodes, (childNode) => {
                 ArrayPush.apply(result, innerTextCollectionSteps(childNode));
             });
+            // node.childNodes.forEach((childNode) => {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts]: dead code

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    if (!nodeIsBeingRendered(thisComputedStyle)) {
+        return this.textContent || ''; // textContentGetterPatched.call(this);
+    }
+
+    const selectionState = getSelectionState(this);
+
+    // 2. Let results be a new empty list.
+    let results: InnerTextCollectionResult[] = [];
+    // 3. For each child node node of this:
+    const childNodes = this.childNodes;
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:200]: Here is where the heart of this implementation is. Add a note here that `this.childNodes` will excludes shadow trees because `Node.childNodes` is patched.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 7]
```diff
@@ -5,8 +5,16 @@
...
+import {
+    hasAttribute,
+    innerTextGetter,
+    innerTextSetter,
+    outerTextGetter,
+    outerTextSetter,
+    tabIndexGetter,
+    tabIndexSetter,
+} from '../env/element';
+import { isNull, isFalse, defineProperties, defineProperty } from '@lwc/shared';
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:17]: Imports from [external modules](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md) should go above local imports. Same for import from `@lwc/features`.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 8]
```diff
@@ -159,6 +171,28 @@ defineProperties(HTMLElement.prototype, {
         writable: true,
         configurable: true,
     },
+    innerText: {
+        get(this: Element): string {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:175]:
```suggestion
        get(this: HTMLElement): string {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 9]
```diff
@@ -159,6 +171,28 @@ defineProperties(HTMLElement.prototype, {
...
+                return innerTextGetter.call(this);
+            }
+
+            // TODO [#1222]: remove global bypass
+            if (isGlobalPatchingSkipped(this)) {
+                return innerTextGetter.call(this);
+            }
+            return getInnerText.call(this);
+        },
+        set(v: string) {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:190]:
```suggestion
        set(this: HTMLElement, v: string) {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 10]
```diff
@@ -170,3 +204,29 @@ defineProperties(HTMLElement.prototype, {
...
+                return outerTextGetter!.call(this);
+            }
+
+            // TODO [#1222]: remove global bypass
+            if (isGlobalPatchingSkipped(this)) {
+                return outerTextGetter!.call(this);
+            }
+            return getInnerText.call(this);
+        },
+        set(v: string) {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:226]:
```suggestion
        set(this: HTMLElement, v: string) {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 11]
```diff
@@ -170,3 +204,29 @@ defineProperties(HTMLElement.prototype, {
         configurable: true,
     },
 });
+
+// Note: Firefox does not have outerText, https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/outerText
+if (outerTextGetter !== null && outerTextSetter !== null) {
+    defineProperty(HTMLElement.prototype, 'outerText', {
+        get(this: Element): string {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:211]:
```suggestion
        get(this: HTMLElement): string {
```

File: packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
+import { createElement } from 'lwc';
+import Container from 'x/container';
+
+if (!process.env.NATIVE_SHADOW) {
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js:4]: Are all tests failing in native shadow? This means the patch is diverging from native behavior widely.

File: packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    const outerTextDescriptor = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'outerText');
+
+    // Firefox does not have outerText.
+    if (outerTextDescriptor) {
+        describe('outerText', () => {
+            const elm = createElement('x-container', { is: Container });
+            document.body.appendChild(elm);
+
+            return Promise.resolve().then(() => {
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js:123]: Don't see a reason to wait for a microtask.
```suggestion
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 14]
```diff
@@ -159,6 +171,28 @@ defineProperties(HTMLElement.prototype, {
...
+            }
+
+            // TODO [#1222]: remove global bypass
+            if (isGlobalPatchingSkipped(this)) {
+                return innerTextGetter.call(this);
+            }
+            return getInnerText.call(this);
+        },
+        set(v: string) {
+            innerTextSetter.call(this, v);
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:191]: This might not be sufficient 🤔 , add tests to make sure it works as expected.

File: packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+                    const testElement = elm.shadowRoot.querySelector(
+                        '.with-slotted-content-2-levels'
+                    );
+
+                    expect(testElement.outerText).toBe(
+                        'first text\n\nslotted element\n\nsecond text'
+                    );
+                });
+            });
+        });
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/synthetic-shadow/inner-outer-text/inner-outer-text.spec.js:148]: Lets add a test to confirm the setter is working as expected:
1. Setting innerText of a host element should
   i. Keep the shadow tree intact
   ii. set only the light dom to the new content
   iii. the light dom is correctly slotted into the shadow tree of the host(in synthetic-mode especially)

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 16]
```diff
@@ -0,0 +1,243 @@
+/*
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:1]: oops, just saw Caridy's suggestion. 🤔  May be leave it here. I do not have a strong opinion against keeping it here. It is just a logical grouping anyway.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+        const tagName = node.tagName;
+        const computedStyle = getElementComputedStyle(node);
+
+        if (tagName === 'OPTION') {
+            // For options, is hard to get the "rendered" text, let's use the original getter.
+            return [1, innerTextGetter.call(node), 1];
+        } else if (tagName === 'TEXTAREA') {
+            return [];
+        } else {
+            const childNodes = node.childNodes;
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:132]: Use patched child node accessor.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    const win = getOwnerWindow(node);
+
+    return windowGetSelection.call(win);
+}
+
+function nodeIsBeingRendered(nodeComputedStyle: CSSStyleDeclaration): boolean {
+    return nodeComputedStyle.visibility === 'visible' && nodeComputedStyle.display !== 'none';
+}
+
+type SelectionState = {
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:32]:
```suggestion
interface SelectionState {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 19]
```diff
@@ -170,3 +204,29 @@ defineProperties(HTMLElement.prototype, {
...
+            }
+
+            // TODO [#1222]: remove global bypass
+            if (isGlobalPatchingSkipped(this)) {
+                return outerTextGetter!.call(this);
+            }
+            return getInnerText.call(this);
+        },
+        set(v: string) {
+            outerTextSetter!.call(this, v);
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:227]: It's an edge case but invoking the `outerText` setter on a host element should trigger its disconnection. Until we merge node reaction, it will not work. 

We can always reimplement the outerText setter in JavaScript ([blink implementation](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/html/html_element.cc;l=841-879;drc=6e8b402a6231405b753919029c9027404325ea00;bpv=0;bpt=1)), but I think it will be a waste of time.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+        result.push(getTextNodeInnerText(node));
+    }
+
+    return result;
+}
+
+/**
+ * innerText spec: https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute
+ */
+export function getInnerText(this: Element): string {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:188]: this should not be expecting `this`, but just a first argument `elm`

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts [hunk 21]
```diff
@@ -170,3 +204,29 @@ defineProperties(HTMLElement.prototype, {
...
     },
 });
+
+// Note: Firefox does not have outerText, https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/outerText
+if (outerTextGetter !== null && outerTextSetter !== null) {
+    defineProperty(HTMLElement.prototype, 'outerText', {
+        get(this: Element): string {
+            if (!featureFlags.ENABLE_ELEMENT_PATCH) {
+                if (isNodeShadowed(this) || isHostElement(this)) {
+                    return getInnerText.call(this);
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:214]: this is outerText, why are you callling getInnerText here? Also, I could not fine the outer text implementation anymore in this diff.

Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:214]: There's no spec for outerText. From [mdn](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/outerText):

> HTMLElement.outerText is a non-standard property. As a getter, it returns the same value as Node.innerText. As a setter, it removes the current node and replaces it with the given text.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+            const childNodes = node.childNodes;
+            for (let i = 0, n = childNodes.length; i < n; i++) {
+                ArrayPush.apply(result, innerTextCollectionSteps(childNodes[i]));
+            }
+        }
+
+        // 2. If node's computed value of 'visibility' is not 'visible', then return items.
+        // 3. If node is not being rendered, then return items. Especial cases: select, datalist, optgroup, option
+        if (!nodeIsBeingRendered(computedStyle)) {
+            if (tagName === 'SELECT' || tagName === 'DATALIST') {
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:141]: the [optgroup](https://html.spec.whatwg.org/multipage/form-elements.html#the-optgroup-element) does not have any text, just options. option is handled in L126. This being said, this is the closest approximation i could came up with from the step in spec, but im open to ideas/suggestions.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+/**
+ * innerText spec: https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute
+ */
+export function getInnerText(this: Element): string {
+    const thisComputedStyle = getElementComputedStyle(this);
+    // 1. If this is not being rendered or if the user agent is a non-CSS user agent, then return this's descendant text content.
+    if (!nodeIsBeingRendered(thisComputedStyle)) {
+        return this.textContent || ''; // textContentGetterPatched.call(this);
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:192]: sure, i thought that moving this implementation to `3rdparty` was to provide a copy-able implementation of the innerText, i would replace the dot notation with the getters (the patched ones).

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts (hunk 14 above)
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:191]: yeah, i haven't started with that, from the pr desc:  _Pending: innerText setter implementation._

**Comment by jodarove** [comment]
@caridy @pmdartus this is ready for another pass, the latest changes were: resolve conflicts and adding the inner/outer text functionality behind a feature flag (`ENABLE_INNER_OUTER_TEXT_PATCH`).

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+interface SelectionState {
+    element: Element;
+    onselect: ((this: GlobalEventHandlers, ev: Event) => any) | null;
+    onselectionchange: ((this: GlobalEventHandlers, ev: Event) => any) | null;
+    onselectstart: ((this: GlobalEventHandlers, ev: Event) => any) | null;
+    ranges: Range[];
+}
+
+function getElementComputedStyle(element: Element): CSSStyleDeclaration {
+    const win = getOwnerWindow(element);
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:26]: I am wondering what is the motivation behind retrieving the element owner window vs. using the current window? Synthetic shadow only applies to the current window.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+    win.onselectstart = onselectstart;
+    win.onselectionchange = onselectionchange;
+}
+
+/**
+ * Gets the "innerText" of a text node using the Selection API
+ *
+ * NOTE: For performance reasons, since this function will be called multiple times while calculating the innerText of
+ *       an element, it does not restore the current selection.
+ * @param textNode
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:94]:
```suggestion
```

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+    return selection.toString();
+}
+
+const nodeIsElement = (node: Node): node is Element => node.nodeType === ELEMENT_NODE;
+const nodeIsText = (node: Node): node is Text => node.nodeType === TEXT_NODE;
+
+/**
+ * Spec: https://html.spec.whatwg.org/multipage/dom.html#inner-text-collection-steps
+ * One spec implementation: https://github.com/servo/servo/blob/721271dcd3c20db5ca8cf146e2b5907647afb4d6/components/layout/query.rs#L1132
+ * @param node
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:125]:
```suggestion
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts (hunk 19 above)
Comment (pmdartus) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:252]: Can we add a comment for this?

**Review by pmdartus — APPROVED** [review]
Added a couple of comments. Otherwise, the code looks good.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts (hunk 24 above)
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:26]: `window.getSelection()` works only for the iframe window, and we rely on it in the polyfill, when calling `.innerText` on an element inside an iframe, we need to get the window for that element.

tbh, I don't know if is required for `getElementComputedStyle`, I used `getOwnerWindow` in this case to follow the same pattern used for the others.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts (hunk 19 above)
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/html-element.ts:255]: done

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+        const { tagName } = node;
+        const computedStyle = getElementComputedStyle(node);
+
+        if (tagName === 'OPTION') {
+            // For options, is hard to get the "rendered" text, let's use the original getter.
+            return [1, innerTextGetter!.call(node), 1];
+        } else if (tagName === 'TEXTAREA') {
+            return [];
+        } else {
+            const childNodes = childNodesGetterPatched.call(node);
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:138]: why are u using this instead of just `node.childNodes`?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/node.ts [hunk 28]
```diff
@@ -164,7 +164,7 @@ function cloneNodePatched(this: Node, deep?: boolean): Node {
 /**
  * This method only applies to elements with a shadow or slots
  */
-function childNodesGetterPatched(this: Node): NodeListOf<Node> {
+export function childNodesGetterPatched(this: Node): NodeListOf<Node> {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/node.ts:167]: this seems problematic to me.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts (hunk 27 above)
Comment (jodarove) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:138]: mainly because of [this comment](https://github.com/salesforce/lwc/pull/2015#discussion_r480130390)

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts (hunk 24 above)
Comment (pmdartus) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:26]: Sounds reasonable to me. I don't see any drawbacks with looking up the owner window on the nodes.

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts (hunk 27 above)
Comment (pmdartus) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:138]: When using the standard `node.childNodes` it is less obvious if you are using the original `Node.prototype.childNodes` or the patched version. I suggested this change because Invoking the patched version directly eliminates this issue.

**Review by caridy — APPROVED** [review]

**Review by caridy — CHANGES_REQUESTED** [review]
wait...

File: packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts (hunk 27 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:138]: well, if that's the case, we might need to do more, consider that when going with the dot notation, it checks if node as `hasMountedChildren`, which translates to is slot or host element, but when calling the patched directly, it checks if it is a host, that seems to be related to my comment in that patched method:

```
    // nothing to do here since this does not have a synthetic shadow attached to it
    // TODO [#1636]: what about slot elements?
```

Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts:138]: in general, something is off!

**Comment by jodarove** [comment]
based on caridy's [comment](https://github.com/salesforce/lwc/pull/2015#discussion_r545324483), i switched back to use the `node.childNodes` notation in b9f4a34. turns out that `childNodesGetterPatched` does not handle properly the `slot` elements, `childNodes` is patched in the [HTMLSlotElement.prototype](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/synthetic-shadow/src/faux-shadow/slot.ts#L177-L190)

@caridy @pmdartus this is ready for another pass 🤞

**Review by pmdartus — APPROVED** [review]

**Comment by pmdartus** [comment]
Did another pass, looks good to me.

**Review by caridy — APPROVED** [review]
